File: src/burp_mcp_lite/upstream.py
"""SSE client to Burp's official MCP server.

Burp exposes the proxy history through the `get_proxy_http_history` tool,
returning paginated results. Each result item is a JSON string of
`{"request": "...", "response": "...", "notes": "..."}`.

We reconnect lazily — there's no benefit to keeping the SSE pipe open between
infrequent refreshes. Each `fetch(offset)` call opens one session, pages
through Burp's history starting at `offset`, and closes the session.
"""

from __future__ import annotations

import json
import logging
from contextlib import AsyncExitStack
from typing import Optional

from mcp import ClientSession
from mcp.client.sse import sse_client

from .snapshot import RawEntry

log = logging.getLogger(__name__)

# A single "page" we ask Burp for. Empirically 20 is reliable on real Burp
# histories; larger pages (e.g. 200) cause the official server to close the
# SSE pipe mid-fetch on entries with very long Cookie headers.
DEFAULT_PAGE = 20


class BurpUpstream:
    """Stateless client. Burp's MCP server runs on 127.0.0.1:9876 by default
    and is bound to localhost; no auth needed.
    """

    def __init__(
        self,
        url: str = "http://127.0.0.1:9876/",
        page_size: int = DEFAULT_PAGE,
        max_entries: int = 5000,
    ) -> None:
        self.url = url
        self.page_size = page_size
        self.max_entries = max_entries

    async def fetch(self, offset: int = 0) -> list[RawEntry]:
        """Fetch the slice of proxy history starting at `offset`, capped at
        `self.max_entries`. Pages internally so a single call can return more
        than `page_size` entries without the caller having to re-loop.
        """
        out: list[RawEntry] = []
        cap = self.max_entries
        async with AsyncExitStack() as stack:
            read, write = await stack.enter_async_context(sse_client(self.url))
            session: ClientSession = await stack.enter_async_context(
                ClientSession(read, write)
            )
            await session.initialize()

            cur_offset = offset
            while len(out) < cap:
                count = min(self.page_size, cap - len(out))
                result = await session.call_tool(
                    "get_proxy_http_history",
                    {"count": count, "offset": cur_offset},
                )
                page = _decode_history_result(result)
                if not page:
                    break
                out.extend(page)
                if len(page) < count:
                    break
                cur_offset += len(page)
        return out


TRUNCATION_MARKER = "... (truncated)"


def _decode_history_result(result) -> list[RawEntry]:
    """Decode the upstream `CallToolResult`.

    The official server's `mcpPaginatedTool` joins all paginated entries
    (each a JSON-encoded `{"request":...,"response":...,"notes":...}`) into a
    single TextContent separated by `\\n\\n`. When the offset is past the end,
    it emits the literal string "Reached end of items".

    Catch: the upstream applies a blunt 5000-char cap per entry
    (`truncateIfNeeded` in Burp's `Tools.kt`) and appends "... (truncated)" to
    the chopped JSON. That makes the JSON unparseable. We salvage what we can
    from the partial string with a tolerant extractor.
    """
    entries: list[RawEntry] = []
    contents = getattr(result, "content", None) or []
    for c in contents:
        text = getattr(c, "text", None)
        if not text:
            continue
        if text == "Reached end of items":
            continue
        if text.startswith("HTTP history access denied"):
            raise UpstreamDenied(text)
        for piece in text.split("\n\n"):
            piece = piece.strip()
            if not piece:
                continue
            entry = _decode_one(piece)
            if entry is not None:
                entries.append(entry)
    return entries


def _decode_one(piece: str) -> Optional[RawEntry]:
    """Parse one entry. Falls back to a tolerant salvage on truncated JSON."""
    try:
        obj = json.loads(piece)
        return (
            obj.get("request") or "",
            obj.get("response") or "",
            obj.get("notes"),
        )
    except json.JSONDecodeError:
        pass

    # Salvage path: drop the truncation marker, then pull the request and
    # response string values out by manual scan.
    body = piece
    if body.endswith(TRUNCATION_MARKER):
        body = body[: -len(TRUNCATION_MARKER)]

    req = _extract_json_string(body, '"request":"')
    resp = _extract_json_string(body, '"response":"')
    notes = _extract_json_string(body, '"notes":"')
    if req is None and resp is None:
        log.warning(
            "upstream returned unparseable entry; skipping (head: %r)", piece[:80]
        )
        return None
    return (req or "", resp or "", notes)


def _extract_json_string(text: str, key: str) -> Optional[str]:
    """Extract a JSON-encoded string value that starts right after `key`.

    `key` includes the opening quote, e.g. ``\"request\":\"``. Decodes JSON
    string escape sequences. If the closing quote is missing (the source was
    truncated mid-string), returns the partial decoded value.
    """
    idx = text.find(key)
    if idx == -1:
        return None
    i = idx + len(key)
    out: list[str] = []
    simple = {"n": "\n", "r": "\r", "t": "\t", "b": "\b", "f": "\f", '"': '"', "\\": "\\", "/": "/"}
    while i < len(text):
        ch = text[i]
        if ch == '"':
            return "".join(out)
        if ch == "\\":
            if i + 1 >= len(text):
                return "".join(out)
            esc = text[i + 1]
            if esc in simple:
                out.append(simple[esc])
                i += 2
                continue
            if esc == "u":
                if i + 6 > len(text):
                    return "".join(out)
                try:
                    out.append(chr(int(text[i + 2 : i + 6], 16)))
                except ValueError:
                    return "".join(out)
                i += 6
                continue
            # Unknown escape — preserve raw backslash + next char.
            out.append(ch)
            i += 1
            continue
        out.append(ch)
        i += 1
    return "".join(out)


class UpstreamDenied(RuntimeError):
    """Burp denied history access (per-call approval, etc.)."""


class FixtureUpstream:
    """Loads pre-recorded history from a JSON fixture file. Used for offline
    development and tests without a live Burp instance.

    File format: a JSON array of objects with keys `request`, `response`,
    optional `notes`.
    """

    def __init__(self, path: str) -> None:
        self.path = path

    async def fetch(self, offset: int = 0) -> list[RawEntry]:
        with open(self.path, "r", encoding="utf-8") as f:
            data = json.load(f)
        all_entries: list[RawEntry] = [
            (obj.get("request") or "", obj.get("response") or "", obj.get("notes"))
            for obj in data
        ]
        return all_entries[offset:]

File: src/upstream.ts
// SSE client to Burp's official MCP server.
//
// Burp exposes proxy history through `get_proxy_http_history`, returning
// paginated results. Each result item is a JSON string of
// {"request": "...", "response": "...", "notes": "..."}.
//
// We reconnect lazily — there's no benefit to keeping the SSE pipe open
// between infrequent refreshes. Each fetch(offset) call opens one session,
// pages through Burp's history starting at offset, and closes the session.

import * as fs from "node:fs/promises";
import { Client } from "@modelcontextprotocol/sdk/client/index.js";
import { SSEClientTransport } from "@modelcontextprotocol/sdk/client/sse.js";
import type { CallToolResult } from "@modelcontextprotocol/sdk/types.js";

import type { RawEntry } from "./snapshot.js";

// A single page we ask Burp for. Empirically 20 is reliable on real Burp
// histories; larger pages cause the official server to close the SSE pipe
// mid-fetch on entries with very long Cookie headers.
export const DEFAULT_PAGE = 20;

const TRUNCATION_MARKER = "... (truncated)";

export class UpstreamDeniedError extends Error {
  constructor(message: string) {
    super(message);
    this.name = "UpstreamDeniedError";
  }
}

export interface Upstream {
  fetch(offset?: number): Promise<RawEntry[]>;
}

export class BurpUpstream implements Upstream {
  readonly url: URL;
  readonly pageSize: number;
  readonly maxEntries: number;

  constructor(opts: {
    url?: string;
    pageSize?: number;
    maxEntries?: number;
  } = {}) {
    this.url = new URL(opts.url ?? "http://127.0.0.1:9876/");
    this.pageSize = opts.pageSize ?? DEFAULT_PAGE;
    this.maxEntries = opts.maxEntries ?? 5000;
  }

  async fetch(offset = 0): Promise<RawEntry[]> {
    const out: RawEntry[] = [];
    const cap = this.maxEntries;
    const transport = new SSEClientTransport(this.url);
    const client = new Client(
      { name: "burp-mcp-lite-upstream", version: "0.2.0" },
      { capabilities: {} },
    );
    try {
      await client.connect(transport);
      let curOffset = offset;
      while (out.length < cap) {
        const count = Math.min(this.pageSize, cap - out.length);
        const result = (await client.callTool({
          name: "get_proxy_http_history",
          arguments: { count, offset: curOffset },
        })) as CallToolResult;
        const page = decodeHistoryResult(result);
        if (page.length === 0) break;
        out.push(...page);
        if (page.length < count) break;
        curOffset += page.length;
      }
    } finally {
      await client.close().catch(() => undefined);
    }
    return out;
  }
}

export class FixtureUpstream implements Upstream {
  readonly path: string;

  constructor(path: string) {
    this.path = path;
  }

  async fetch(offset = 0): Promise<RawEntry[]> {
    const text = await fs.readFile(this.path, "utf-8");
    const data = JSON.parse(text) as Array<{
      request?: string;
      response?: string;
      notes?: string | null;
    }>;
    const all: RawEntry[] = data.map((obj) => ({
      request: obj.request ?? "",
      response: obj.response ?? "",
      notes: obj.notes ?? null,
    }));
    return all.slice(offset);
  }
}

export function decodeHistoryResult(result: CallToolResult): RawEntry[] {
  const entries: RawEntry[] = [];
  const contents = result.content ?? [];
  for (const c of contents) {
    if (c.type !== "text") continue;
    const text = (c as { text?: string }).text ?? "";
    if (!text) continue;
    if (text === "Reached end of items") continue;
    if (text.startsWith("HTTP history access denied")) {
      throw new UpstreamDeniedError(text);
    }
    for (const piece of text.split("\n\n")) {
      const trimmed = piece.trim();
      if (!trimmed) continue;
      const entry = decodeOne(trimmed);
      if (entry) entries.push(entry);
    }
  }
  return entries;
}

function decodeOne(piece: string): RawEntry | null {
  // Try strict JSON first.
  try {
    const obj = JSON.parse(piece) as {
      request?: string;
      response?: string;
      notes?: string | null;
    };
    return {
      request: obj.request ?? "",
      response: obj.response ?? "",
      notes: obj.notes ?? null,
    };
  } catch {
    // fall through to salvage path
  }

  // Salvage: drop the truncation marker, then pull request/response/notes
  // string values out by manual scan. Burp's Tools.kt caps each entry at
  // 5000 chars; if truncation cut us mid-string-mid-escape, the JSON parse
  // above failed but we can still recover the partial values.
  let body = piece;
  if (body.endsWith(TRUNCATION_MARKER)) {
    body = body.slice(0, -TRUNCATION_MARKER.length);
  }

  const req = extractJsonString(body, '"request":"');
  const resp = extractJsonString(body, '"response":"');
  const notes = extractJsonString(body, '"notes":"');
  if (req === null && resp === null) return null;
  return {
    request: req ?? "",
    response: resp ?? "",
    notes,
  };
}

function extractJsonString(text: string, key: string): string | null {
  const idx = text.indexOf(key);
  if (idx === -1) return null;
  let i = idx + key.length;
  const out: string[] = [];
  const simple: Record<string, string> = {
    n: "\n",
    r: "\r",
    t: "\t",
    b: "\b",
    f: "\f",
    '"': '"',
    "\\": "\\",
    "/": "/",
  };
  while (i < text.length) {
    const ch = text[i]!;
    if (ch === '"') return out.join("");
    if (ch === "\\") {
      if (i + 1 >= text.length) return out.join("");
      const esc = text[i + 1]!;
      if (esc in simple) {
        out.push(simple[esc]!);
        i += 2;
        continue;
      }
      if (esc === "u") {
        if (i + 6 > text.length) return out.join("");
        const cp = Number.parseInt(text.slice(i + 2, i + 6), 16);
        if (!Number.isFinite(cp)) return out.join("");
        out.push(String.fromCharCode(cp));
        i += 6;
        continue;
      }
      out.push(ch);
      i += 1;
      continue;
    }
    out.push(ch);
    i += 1;
  }
  return out.join("");
}
